File: tests/legacy_vectors/cross-sdk/run_tests.js
#!/usr/bin/env node
/**
 * Cross-SDK Test Vector Runner for Node.js
 *
 * Runs the standard ASH test vectors against the Node.js SDK to verify
 * interoperability with other language implementations.
 *
 * Usage:
 *   node run_tests.js
 *   node run_tests.js --verbose
 *   node run_tests.js --category canonicalization
 */

const fs = require('fs');
const path = require('path');

// Try to load the SDK
let ash;
try {
  ash = require('../../packages/ash-node-sdk/dist/index.js');
} catch (e) {
  console.log('Warning: Could not load compiled SDK, trying source...');
  try {
    // This would require ts-node or pre-compilation
    console.log('SDK not available. Build first: cd packages/ash-node-sdk && npm run build');
    process.exit(1);
  } catch (e2) {
    console.error('ERROR: ASH Node.js SDK not available');
    console.error('Build with: cd packages/ash-node-sdk && npm run build');
    process.exit(1);
  }
}

function loadTestVectors() {
  const vectorsPath = path.join(__dirname, 'test-vectors.json');
  return JSON.parse(fs.readFileSync(vectorsPath, 'utf8'));
}

function runCanonicalizationTests(vectors, verbose) {
  console.log('\n=== JSON Canonicalization Tests ===');
  let passed = 0;
  let failed = 0;

  for (const test of vectors.canonicalization.vectors) {
    try {
      const result = ash.ashCanonicalizeJson(test.input);
      if (result === test.expected) {
        passed++;
        if (verbose) {
          console.log(`  [PASS] ${test.id}: ${test.description}`);
        }
      } else {
        failed++;
        console.log(`  [FAIL] ${test.id}: ${test.description}`);
        console.log(`    Expected: ${test.expected}`);
        console.log(`    Got:      ${result}`);
      }
    } catch (e) {
      failed++;
      console.log(`  [FAIL] ${test.id}: ${test.description}`);
      console.log(`    Error: ${e.message}`);
    }
  }

  console.log(`\nCanonicalization: ${passed} passed, ${failed} failed`);
  return { passed, failed };
}

function runUrlencodedTests(vectors, verbose) {
  console.log('\n=== URL-Encoded Canonicalization Tests ===');
  let passed = 0;
  let failed = 0;

  for (const test of vectors.urlencoded_canonicalization.vectors) {
    try {
      const result = ash.ashCanonicalizeUrlencoded(test.input);
      if (result === test.expected) {
        passed++;
        if (verbose) {
          console.log(`  [PASS] ${test.id}: ${test.description}`);
        }
      } else {
        failed++;
        console.log(`  [FAIL] ${test.id}: ${test.description}`);
        console.log(`    Expected: ${test.expected}`);
        console.log(`    Got:      ${result}`);
      }
    } catch (e) {
      failed++;
      console.log(`  [FAIL] ${test.id}: ${test.description}`);
      console.log(`    Error: ${e.message}`);
    }
  }

  console.log(`\nURL-Encoded: ${passed} passed, ${failed} failed`);
  return { passed, failed };
}

function runBindingTests(vectors, verbose) {
  console.log('\n=== Binding Normalization Tests ===');
  let passed = 0;
  let failed = 0;

  for (const test of vectors.binding_normalization.vectors) {
    try {
      const result = ash.ashNormalizeBinding(test.method, test.path, test.query);
      if (result === test.expected) {
        passed++;
        if (verbose) {
          console.log(`  [PASS] ${test.id}: ${test.description}`);
        }
      } else {
        failed++;
        console.log(`  [FAIL] ${test.id}: ${test.description}`);
        console.log(`    Expected: ${test.expected}`);
        console.log(`    Got:      ${result}`);
      }
    } catch (e) {
      failed++;
      console.log(`  [FAIL] ${test.id}: ${test.description}`);
      console.log(`    Error: ${e.message}`);
    }
  }

  console.log(`\nBinding: ${passed} passed, ${failed} failed`);
  return { passed, failed };
}

function runTimingSafeTests(vectors, verbose) {
  console.log('\n=== Timing-Safe Comparison Tests ===');
  let passed = 0;
  let failed = 0;

  for (const test of vectors.timing_safe_equal.vectors) {
    try {
      const result = ash.ashTimingSafeEqual(test.a, test.b);
      if (result === test.expected) {
        passed++;
        if (verbose) {
          console.log(`  [PASS] ${test.id}: a="${test.a}", b="${test.b}"`);
        }
      } else {
        failed++;
        console.log(`  [FAIL] ${test.id}: a="${test.a}", b="${test.b}"`);
        console.log(`    Expected: ${test.expected}`);
        console.log(`    Got:      ${result}`);
      }
    } catch (e) {
      failed++;
      console.log(`  [FAIL] ${test.id}`);
      console.log(`    Error: ${e.message}`);
    }
  }

  console.log(`\nTiming-Safe: ${passed} passed, ${failed} failed`);
  return { passed, failed };
}

function main() {
  const args = process.argv.slice(2);
  const verbose = args.includes('--verbose') || args.includes('-v');
  const categoryIndex = args.findIndex(a => a === '--category' || a === '-c');
  const category = categoryIndex >= 0 ? args[categoryIndex + 1] : null;

  console.log('='.repeat(60));
  console.log('ASH Cross-SDK Test Vector Runner - Node.js');
  console.log('='.repeat(60));

  const vectors = loadTestVectors();
  let totalPassed = 0;
  let totalFailed = 0;

  const categories = {
    canonicalization: runCanonicalizationTests,
    urlencoded: runUrlencodedTests,
    binding: runBindingTests,
    timing: runTimingSafeTests,
  };

  if (category) {
    if (categories[category]) {
      const { passed, failed } = categories[category](vectors, verbose);
      totalPassed += passed;
      totalFailed += failed;
    } else {
      console.log(`Unknown category: ${category}`);
      console.log(`Available: ${Object.keys(categories).join(', ')}`);
      process.exit(1);
    }
  } else {
    for (const [name, func] of Object.entries(categories)) {
      const { passed, failed } = func(vectors, verbose);
      totalPassed += passed;
      totalFailed += failed;
    }
  }

  console.log('\n' + '='.repeat(60));
  console.log(`TOTAL: ${totalPassed} passed, ${totalFailed} failed`);
  console.log('='.repeat(60));

  process.exit(totalFailed === 0 ? 0 : 1);
}

main();

File: tests/legacy_vectors/cross-sdk/run_tests.py
#!/usr/bin/env python3
"""
Cross-SDK Test Vector Runner for Python

Runs the standard ASH test vectors against the Python SDK to verify
interoperability with other language implementations.

Usage:
    python run_tests.py
    python run_tests.py --verbose
    python run_tests.py --category canonicalization
"""

import json
import sys
import argparse
from pathlib import Path

# Add the SDK to path
SDK_PATH = Path(__file__).parent.parent.parent / "packages" / "ash-python-sdk" / "src"
sys.path.insert(0, str(SDK_PATH))

try:
    from ash.core import (
        canonicalize_json,
        canonicalize_url_encoded,
        normalize_binding,
        hash_body,
        timing_safe_compare,
    )
    SDK_AVAILABLE = True
except ImportError as e:
    print(f"Warning: Could not import ash SDK: {e}")
    SDK_AVAILABLE = False


def load_test_vectors():
    """Load test vectors from JSON file."""
    vectors_path = Path(__file__).parent / "test-vectors.json"
    with open(vectors_path, "r", encoding="utf-8") as f:
        return json.load(f)


def run_canonicalization_tests(vectors, verbose=False):
    """Run JSON canonicalization test vectors."""
    print("\n=== JSON Canonicalization Tests ===")
    passed = 0
    failed = 0

    for test in vectors["canonicalization"]["vectors"]:
        try:
            # Parse input JSON string to Python object first
            input_obj = json.loads(test["input"])
            result = canonicalize_json(input_obj)
            if result == test["expected"]:
                passed += 1
                if verbose:
                    print(f"  [PASS] {test['id']}: {test['description']}")
            else:
                failed += 1
                print(f"  [FAIL] {test['id']}: {test['description']}")
                print(f"    Expected: {test['expected']}")
                print(f"    Got:      {result}")
        except Exception as e:
            failed += 1
            print(f"  [FAIL] {test['id']}: {test['description']}")
            print(f"    Error: {e}")

    print(f"\nCanonicalisation: {passed} passed, {failed} failed")
    return passed, failed


def run_urlencoded_tests(vectors, verbose=False):
    """Run URL-encoded canonicalization test vectors."""
    print("\n=== URL-Encoded Canonicalization Tests ===")
    passed = 0
    failed = 0

    for test in vectors["urlencoded_canonicalization"]["vectors"]:
        try:
            result = canonicalize_url_encoded(test["input"])
            if result == test["expected"]:
                passed += 1
                if verbose:
                    print(f"  [PASS] {test['id']}: {test['description']}")
            else:
                failed += 1
                print(f"  [FAIL] {test['id']}: {test['description']}")
                print(f"    Expected: {test['expected']}")
                print(f"    Got:      {result}")
        except Exception as e:
            failed += 1
            print(f"  [FAIL] {test['id']}: {test['description']}")
            print(f"    Error: {e}")

    print(f"\nURL-Encoded: {passed} passed, {failed} failed")
    return passed, failed


def run_binding_tests(vectors, verbose=False):
    """Run binding normalization test vectors."""
    print("\n=== Binding Normalization Tests ===")
    passed = 0
    failed = 0

    for test in vectors["binding_normalization"]["vectors"]:
        try:
            result = normalize_binding(test["method"], test["path"], test["query"])
            if result == test["expected"]:
                passed += 1
                if verbose:
                    print(f"  [PASS] {test['id']}: {test['description']}")
            else:
                failed += 1
                print(f"  [FAIL] {test['id']}: {test['description']}")
                print(f"    Expected: {test['expected']}")
                print(f"    Got:      {result}")
        except Exception as e:
            failed += 1
            print(f"  [FAIL] {test['id']}: {test['description']}")
            print(f"    Error: {e}")

    print(f"\nBinding: {passed} passed, {failed} failed")
    return passed, failed


def run_timing_safe_tests(vectors, verbose=False):
    """Run timing-safe comparison test vectors."""
    print("\n=== Timing-Safe Comparison Tests ===")
    passed = 0
    failed = 0

    for test in vectors["timing_safe_equal"]["vectors"]:
        try:
            result = timing_safe_compare(test["a"], test["b"])
            if result == test["expected"]:
                passed += 1
                if verbose:
                    print(f"  [PASS] {test['id']}: a={repr(test['a'])}, b={repr(test['b'])}")
            else:
                failed += 1
                print(f"  [FAIL] {test['id']}: a={repr(test['a'])}, b={repr(test['b'])}")
                print(f"    Expected: {test['expected']}")
                print(f"    Got:      {result}")
        except Exception as e:
            failed += 1
            print(f"  [FAIL] {test['id']}")
            print(f"    Error: {e}")

    print(f"\nTiming-Safe: {passed} passed, {failed} failed")
    return passed, failed


def main():
    parser = argparse.ArgumentParser(description="Run ASH cross-SDK test vectors")
    parser.add_argument("--verbose", "-v", action="store_true", help="Show all test results")
    parser.add_argument("--category", "-c", help="Run specific category only")
    args = parser.parse_args()

    if not SDK_AVAILABLE:
        print("ERROR: ASH Python SDK not available")
        print("Install with: pip install -e packages/ash-python-sdk")
        sys.exit(1)

    print("=" * 60)
    print("ASH Cross-SDK Test Vector Runner - Python")
    print("=" * 60)

    vectors = load_test_vectors()
    total_passed = 0
    total_failed = 0

    categories = {
        "canonicalization": run_canonicalization_tests,
        "urlencoded": run_urlencoded_tests,
        "binding": run_binding_tests,
        "timing": run_timing_safe_tests,
    }

    if args.category:
        if args.category in categories:
            p, f = categories[args.category](vectors, args.verbose)
            total_passed += p
            total_failed += f
        else:
            print(f"Unknown category: {args.category}")
            print(f"Available: {', '.join(categories.keys())}")
            sys.exit(1)
    else:
        for name, func in categories.items():
            p, f = func(vectors, args.verbose)
            total_passed += p
            total_failed += f

    print("\n" + "=" * 60)
    print(f"TOTAL: {total_passed} passed, {total_failed} failed")
    print("=" * 60)

    sys.exit(0 if total_failed == 0 else 1)


if __name__ == "__main__":
    main()
